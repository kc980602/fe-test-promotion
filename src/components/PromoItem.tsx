import _ from 'lodash'
import React, { FC } from 'react'
import { Link } from 'react-router-dom'
import AspectRatioBox from '~/components/AspectRatioBox'
import { ThemeColor } from '~/types/app.type'
import { Promotion } from '~/types/promotion.type'
import { lineClamp } from '~/utils/cssMixins'

type BoxImageProps = {
  url: string
}

const BoxImage: FC<BoxImageProps> = ({ url }) => (
  <AspectRatioBox ratio={1}>
    <img src={url} sx={{ objectFit: 'contain' }} />
  </AspectRatioBox>
)

type PromoItemProps = {
  promotion: Promotion
  color?: ThemeColor
}

const PromoItem: FC<PromoItemProps> = ({ promotion, color }) => {

  const imgCount = promotion.productImageUrls.length
  const grid = imgCount > 5 ? 3 : imgCount > 1 ? 2 : 1

  return (
    <a href={promotion.promoUrl} target="_blank">
      <div
        sx={{
          borderWidth: 2,
          borderStyle: 'solid',
          borderColor: `${color}.base`,
          borderRadius: 3
        }}
      >
        <div
          sx={{
            display: 'grid',
            gridGap: 2,
            gridTemplateColumns: `repeat(${grid}, 1fr)`,
            gridTemplateRow: `repeat(${grid}, 1fr)`,
            p: 2
          }}
        >
          {
            _.isEmpty(promotion.productImageUrls) ?
              <BoxImage
                url='https://dummyimage.com/500x500/333333/fff.jpg&text=Placeholder'
              /> :
              promotion.productImageUrls.map((url, idx) => (
                <BoxImage key={idx} url={url} />
              ))
          }
          {/**/}
        </div>
        <div
          sx={{
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            height: 68,
            backgroundColor: `${color}.base`,
            color: 'white',
            fontSize: 18,
            textAlign: 'center',
            p: 1,
          }}
        >
        <span sx={lineClamp(2)}>
          {promotion.promoName}
        </span>
        </div>
      </div>
    </a>
  )
}

export default React.memo(PromoItem)
