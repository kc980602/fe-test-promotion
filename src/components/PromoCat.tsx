import _ from 'lodash'
import React, { FC, useEffect, useState } from 'react'
import PromoItem from '~/components/PromoItem'
import { getPromotions } from '~/slices/promotionSlice'
import { useDispatch, useSelector } from '~/store'
import { ThemeColor } from '~/types/app.type'
import { Category } from '~/types/promotion.type'

type PromoCatProps = {
  categories: Category[],
  color?: ThemeColor
}

const PromoCat: FC<PromoCatProps> = ({ categories, color = 'orange' }) => {
  const [selected, setSelected] = useState<number>()

  const dispatch = useDispatch()

  const { promotions } = useSelector(state => state.promotion)

  useEffect(() => {
    if (categories && !_.isEmpty(categories)) {
      setSelected(categories[0].id)
    }
  }, [categories])

  useEffect(() => {
    // Dispatching the thunk returns a promise
    const promise = !_.isNil(selected) && dispatch(getPromotions(selected))
    return () => {
      // `createAsyncThunk` attaches an `abort()` method to the promise
      promise && promise?.abort()
    }
  }, [selected])

  return (
    <div className='PromoCat'>
      <div
        sx={{
          display: 'flex',
          alignItems: 'baseline'
        }}
      >
        {
          categories.map(item => (
            <button
              key={item.id}
              onClick={() => setSelected(item.id)}
              sx={{
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                width: 200,
                height: 60,
                fontSize: 24,
                p: 1,
                backgroundColor: 'white',
                color: `${color}.base`,
                transition: 'background-color .1s ease-in-out',
                ...(selected === item.id ? {
                  cursor: 'default',
                  height: 64,
                  borderTopWidth: 4,
                  borderTopStyle: 'solid',
                  borderTopColor: `${color}.accent`,
                  backgroundColor: `${color}.base`,
                  color: 'white',
                  boxShadow: '2px 0 4px #a5a5a5',
                  zIndex: 2
                } : {})
              }}
            >
              <span>{item.name}</span>
            </button>
          ))
        }


      </div>
      <div
        sx={{
          position: 'relative',
          zIndex: 2,
          borderWidth: 10,
          borderStyle: 'solid',
          borderColor: `${color}.base`,
          backgroundColor: 'white',
          p: [24, null, 32]
        }}
      >
        <div
          sx={{
            display: ['flex', null, 'grid'],
            gridTemplateColumns: 'repeat(5, minmax(0, 1fr))',
            alignItems: 'center',
            gap: 16,
            overflow: 'auto'
          }}
        >
          {
            promotions.map(item => (
              <div
                key={item.promoUrl}
                sx={{ width: [210, null, 'auto'], flexShrink: 0 }}
              >
                <PromoItem color={color} promotion={item} />
              </div>
            ))
          }
        </div>
      </div>
    </div>
  )
}

export default React.memo(PromoCat)
