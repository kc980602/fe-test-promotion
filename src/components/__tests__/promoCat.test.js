import {shallow} from 'enzyme'
import toJson from 'enzyme-to-json'
import {Provider as ReduxProvider} from 'react-redux'
import configureStore from 'redux-mock-store'
import PromoCat from '~/components/PromoCat'
import PromoItem from '~/components/PromoItem'
import Categories from '../../../public/data/categories.json'
import Content1 from '../../../public/data/content-1.json'


const mockStore = configureStore()
const initialState = {
  promotion: {
    promotions: Content1
  }
}

const store = mockStore(initialState)

describe('PromoCat', () => {
  test('test if promoCat renders', () => {
    const wrapper = shallow(
      <ReduxProvider store={store}>
        <PromoCat categories={Categories} />
      </ReduxProvider>
    )

    const component = wrapper.dive()

    expect(toJson(component)).toMatchSnapshot()
  })

  test('test if promoItem renders', () => {
    const wrapper = shallow(
      <PromoItem promotion={Content1[0]} />
    )

    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
