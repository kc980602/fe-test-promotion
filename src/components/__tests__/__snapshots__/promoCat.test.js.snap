// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PromoCat test if promoCat renders 1`] = `
<Memo(PromoCat)
  categories={
    Array [
      Object {
        "id": 1,
        "name": "Best Offer",
        "url": "/promo-detail?from=best_offer",
      },
      Object {
        "id": 2,
        "name": "New Arrivial",
        "url": "/promo-detail?from=new_arrivial",
      },
      Object {
        "id": 3,
        "name": "Flash Sale",
        "url": "/promo-detail?from=flash_sale",
      },
    ]
  }
/>
`;

exports[`PromoCat test if promoItem renders 1`] = `
<Link
  to="/promo-detail?from=best_offer_1"
>
  <div
    sx={
      Object {
        "borderColor": "undefined.base",
        "borderRadius": 3,
        "borderStyle": "solid",
        "borderWidth": 2,
      }
    }
  >
    <div
      sx={
        Object {
          "display": "grid",
          "gridGap": 2,
          "gridTemplateColumns": "repeat(1, 1fr)",
          "gridTemplateRow": "repeat(1, 1fr)",
          "p": 2,
        }
      }
    >
      <BoxImage
        key="0"
        url="https://dummyimage.com/500x500/86d4f8/fff.jpg&text=A"
      />
    </div>
    <div
      sx={
        Object {
          "alignItems": "center",
          "backgroundColor": "undefined.base",
          "color": "white",
          "display": "flex",
          "fontSize": 18,
          "height": 68,
          "justifyContent": "center",
          "p": 1,
          "textAlign": "center",
        }
      }
    >
      <span
        sx={
          Object {
            "WebkitBoxOrient": "vertical",
            "WebkitLineClamp": 2,
            "display": "-webkit-box",
            "overflow": "hidden",
            "textOverflow": "ellipsis",
            "wordBreak": "break-all",
          }
        }
      >
        Best Offer 1
      </span>
    </div>
  </div>
</Link>
`;
